src/grift/type-check.rkt

```diff
@@ -419,127 +649,221 @@ The type rules for core forms that have interesting type rules
+             (values (Mvector-ref e1 e2) ty))]
+          [(Mvector-set! (app recur e1 t1) (app recur e2 t2) (app recur e3 t3))
+           (let ([ty (mvector-set!-type-rule t1 t2 t3)])
+             (values (Mvector-set! e1 e2 e3) ty))]
+          [(Mvector-length (app recur e t))
+           (values (Mvector-length e) (mvector-length-type-rule t))]
+          [(Gvector (app recur e1 t1) (app recur e2 t2))
+           (values (Gvector e1 e2) (gvector-type-rule t1 t2))]
+          [(Gvector-ref (app recur e1 t1) (app recur e2 t2))
+           (values (Gvector-ref e1 e2) (gvector-ref-type-rule t1 t2))]
+          [(Gvector-set! (app recur e1 t1) (app recur e2 t2) (app recur e3 t3))
+           (values (Gvector-set! e1 e2 e3) (gvector-set!-type-rule t1 t2 t3))]
+          [(Gvector-length (app recur e t))
+           (values (Gvector-length e)
+                   (gvector-length-type-rule t))]
+          [(Create-tuple (app map-recur e* t*))
+           (values (Create-tuple e*) (tuple-type-rule t*))]
+          [(Tuple-proj (app recur e t) i)
+           (values (Tuple-proj e i) (tuple-proj-type-rule t i))]
+          [other (error 'type-check "unmatched ~a" other)]))
+      )
+    (values (Ann new-exp (cons src type)) type))
+  (recur exp))
 
 
 ;; Type checks the rhs to be consistent with type annotation if
 ;; provided the resulting type is the type of the annotation.
-(: mk-tc-binding :
-   Src (-> S0-Expr (values S1-Expr Grift-Type)) -> (-> S0-Bnd S1-Bnd))
+(: tc-binding :
+   Src (-> G0-Ann-Expr (values G1-Ann-Expr Grift-Type)) -> (-> G0-Bnd G1-Bnd))
 (define ((tc-binding src tc-expr) bnd)
-  (match-define (Bnd id id-type rhs) bnd)
+  (match-define (Bnd id t0 rhs) bnd)
   (define-values (new-rhs rhs-type) (tc-expr rhs))
-  (Bnd id (let-binding-type-rule id-type rhs-type id src) new-rhs))
-
-
-
+  (define t1 (let-binding-type-rule (and t0 (validate-type src t0)) rhs-type id src))
+  (Bnd id t1 new-rhs))
+
+(: map-tc-binding :
+   Src (-> G0-Ann-Expr (values G1-Ann-Expr Grift-Type))
+   -> (-> G0-Bnd* G1-Bnd*))
+(define ((map-tc-binding src tc-expr) bnd*)
+  (map (tc-binding src tc-expr) bnd*))
+
+
+(: map-switch-case2 :
+   (All (A B C)
+        (A -> (Values B C))
+        (Switch-Case A) ->
+        (Values (Switch-Case B) C)))
+(define (map-switch-case2 f c)
+  (define-values (r0 r1) (f (cdr c)))
+  (values (cons (car c) r0) r1))
+(: map-switch-case2* :
+   (All (A B C)
+        (A -> (Values B C))
+        (Switch-Case* A)
+        ->
+        (Values (Switch-Case* B) (Listof C))))
+(define (map-switch-case2* f a*)
+  (for/lists ([b* : (Switch-Case* B)]
+              [c* : (Listof C)])
+             ([a a*])
+    (map-switch-case2 f a)))
+
+
+;; resolve :: takes a type and converts it to a equivalent type
+;; where the outermost type is a concrete type constructor (like ->, Tuple, etc.) as
+;; opposed to logical (like Rec).
+;; NOTE: this should only be called on types that have been converted to normal form. 
+(define (resolve t)
```

Comment:
I am not sure if `resolve` is a good name for this function.

Reply:
Perhaps `concretize-type` or `realize-type` would work. It is called `resolve` because that is what typed-racket's function that does the same thing is called.